File: rs/hang/src/catalog/location.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::model::Position;

/// Live location information for spatial audio positioning.
///
/// This struct describes the spatial position of a broadcaster or viewer
/// for implementing spatial audio features like panning and 3D positioning.
///
/// It also contains a mechanism for peers to request position updates.
/// It's up to the broadcaster to fetch the catalog of other peers, detect if their handle is present, and subscribe to those updates.
#[serde_with::serde_as]
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Default)]
#[serde(default, rename_all = "camelCase")]
#[serde_with::skip_serializing_none]
pub struct Location {
	/// Our initial location.
	///
	/// This is the location that we start with. The range is -1 to 1 with (0,0) being the center.
	pub initial: Option<Position>,

	/// An optional track that contains live location updates.
	pub updates: Option<moq_lite::Track>,

	/// If present, this broadcaster is requesting that other peers update their position via the handle.
	#[serde(default, skip_serializing_if = "HashMap::is_empty")]
	pub peers: HashMap<u32, moq_lite::Track>,

	/// If present, viewers can drag this location via this handle
	pub handle: Option<u32>,
}

File: js/hang/src/catalog/location.ts
import { z } from "zod";
import { TrackSchema } from "./track";

export const PositionSchema = z.object({
	// The relative X position of the broadcast, from -1 to +1.
	// This should be used for audio panning but can also be used for video positioning.
	x: z.number().optional(),

	// The relative Y position of the broadcast, from -1 to +1.
	// This can be used for video positioning, and maybe audio panning.
	y: z.number().optional(),

	// The relative Z index of the broadcast, where larger values are closer to the viewer.
	// This is used to break ties when there are multiple broadcasts at the same position.
	z: z.number().optional(),

	// The scale of the broadcast, where 1 is 100%
	s: z.number().optional(),
});

export const LocationSchema = z.object({
	// The initial position of the broadcaster, from -1 to +1 in both dimensions.
	// If not provided, then the broadcaster is assumed to be at (0,0)
	// This should be used for audio panning but can also be used for video positioning.
	initial: PositionSchema.optional(),

	// If provided, then updates to the position are done via a separate Moq track.
	// This is used to avoid a full catalog update every time we want to update a few bytes.
	// TODO: These updates currently use JSON for simplicity, but we should use a binary format.
	track: TrackSchema.optional(),

	// If set, then this broadcaster allows other peers to request position updates via this handle.
	// We will have to discover and subscribe to their position updates.
	handle: z.string().optional(),

	// If provided, this broadcaster is signaling the location of other peers.
	// The payload is a JSON blob keyed by handle for each peer.
	peers: TrackSchema.optional(),
});

export type Location = z.infer<typeof LocationSchema>;
export type Position = z.infer<typeof PositionSchema>;

export const PeersSchema = z.record(z.string(), PositionSchema);
export type Peers = z.infer<typeof PeersSchema>;
